app/models/guest_device.rb

```diff
@@ -24,4 +24,8 @@ def self.with_ethernet_type
   def self.with_storage_type
     where(:device_type => "storage")
   end
+
+  def self.display_name(number = 1)
+    n_('Guest Device', 'Guest Devices', number)
```

Comment:
A large number of these are basically:

```ruby
def self.display_name(number = 1)
  n_(name.titleize, name.titleize.pluralize, number)
end
```

Can we include this method in a mixin added to all models and override this method in the model where they're different?

Reply:
Strictly speaking, this PR just moves those strings from `locale/en.yml` into particular models. So those strings and the fallback mechanism are already in our application, but implemented differently. Reasons why I'm changing it are explained in the initial comment.

Reply:
I still don't understand why we need to override this method in all of these models that hardcode what the ApplicationRecord method does generically, other than where we have special rules, such as MiqReport.

I'm still not sure if burying this method in models is the right place.  Ultimately, are we thinking each model would have a UI presenter type mixin so provider/core developers know what things they need to implement to correctly be presentable and translatable?

Reply:
As things are, we cannot just override special cases like `MiqReport` and have everything else done implicitly, because we need to have those pretty display names translated.

This does not really differ from what we currently have in `locale/en.yml`. The difference is only in implementation: currently it's one yaml file for all models (including the ones in provider repos), what I'm aiming for here is pluggability of the whole concept.

Reply:
> what I'm aiming for here is pluggability of the whole concept.

`locale/en.yml` is not pluggable

This PR is about fixing that. Not about changing names for things.

Reply:
> Can we include this method in a mixin added to all models and override this method in the model where they're different?

You need to put the strings somewhere. Somewhere pluggable. Like in the models.

You cannot have a single implementation with `  n_(name.titleize, name.titleize.pluralize, number)`. Because that code does not contain the strings.

Reply:
> Because that code does not contain the strings that would gettext collect.

I guess this is the part I don't understand.  Is gettext using sed to find the display_name method and return the next line?  Or, is it calling `Model.display_name`.  If it's the latter, I don't understand why we need to hardcode strings that are easily replaced with the conventional naming structure for most models.

Either way, please explain the why we need to duplicate strings that are available programmatically, if it's a limitation in gettext or not.  It's very much unclear in the commit or description.  Perhaps, it's obvious if you use gettext but some of us only deal with gettext when it's bloating worker memory and haven't actually used it for translating things.

At the very least, if we need to duplicate things because of gettext, we should probably include them as a separate mixin so it's more obvious what model writers need to do in order to make their models presentable in the UI.  We don't need to do this now but hopefully, that's the goal.

Reply:
Sorry, it feels like there's something obvious I'm not undestanding

Reply:
I guess we could say this is a limitation of gettext. If there's a string in your application that needs / has to be translated, you just wrap it with gettext: `_("My String.")`. During string extraction (i.e. the process which collects all the strings to translate) only static strings inside gettext these functions make it to the catalog.

Reply:
Thanks.  That makes sense.  Ok, so we're trading duplication of logic so we don't need to dynamically generate these strings to feed to `gettext`.  Right?

I'm ok with that.

Reply:
I think there is some middle ground to be had here.  I agree with @jrafanie that forcing every single model to have a display_name method is kind of overkill.  This puts an onus on the developer and the reviewer to remember that every model requires one of these methods (on top of everything else the reviewer must remember).

Pure coding wise, I think we all agree that a method in ApplicationRecord for the 80% case should be sufficient (e.g. `n_(name.titleize, name.titleize.pluralize, number)`).  This is typical Ruby, and preferable.  I also understand that due to "deficiencies" in the extractor, static strings are required, and "dynamic" code like this is hard, so it makes us code in an inefficient manner.

So, what if we took a middle ground and adopted some of the extraction tricks we've already done for en.yml: https://github.com/ManageIQ/manageiq/blob/master/lib/tasks/locale.rake#L1  That is, created something that enumerates all of the models, calls display_name on them, and stores that result in a Ruby file for use with gettext?